src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,507 @@
+
+
+#### Understanding migration results
+
+To see the reasons for each change (or non-change),
+click its line number in the **Proposed Edits** pane.
+The reasons appear in the **Edit Details** pane.
+
+For example, consider the following code,
+from before null safety:
+
+```dart
+var ints = const <int>[0, null];
+var zero = ints[0];
+var one = zero + 1;
+var zeroOne = <int>[zero, one];
+```
+
+The default migration is backward compatible but not ideal:
+
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0];
+var one = zero! + 1;
+var zeroOne = <int?>[zero, one];
+```
+
+By clicking line 3, you can see the migration tool's reasons for
+adding the `!`.
+But if you know that `zero` can't be null,
+then you can improve the migration result.
+
+
+#### Improving migration results
+
+When analysis infers the wrong nullability,
+you can override its proposed edits by inserting temporary
+`/*?*/` and `/*!*/` markers:
+
+* In the **Edit Details** pane of the migration tool,
+  you can insert these markers using the
+  **Add `/*?*/` hint** and **Add `/*!*/` hint** buttons.
+
+  These buttons add comments to your file immediately,
+  and there's no **Undo**.
+  If you don't want a hint that the tool inserted,
+  you can use your usual code editor to remove it.
+
+* You can use an editor to add these markers,
+  even while the tool is still running.
+  Because your code hasn't opted into null safety yet,
+  you can't use new null-safety features.
+  You can, however, make changes like refactoring
+  that don't depend on null-safety features.
+
+  When you've finished editing your code,
+  click **Rerun from sources** to pick up your changes.
+
+A single hint can have ripple effects elsewhere in the code.
+In the example from before,
+adding a `/*!*/` marker where `zero` is assigned its value (on line 2)
+makes the migration tool infer the type of `zero` as `int?` instead of `int`.
+This type change can affect code that directly or indirectly uses `zero`.
+
+```dart
+var zero = ints[0]/*!*/;
+```
+
+With the above hint, the migration tool changes its proposed edits,
+as the following table shows.
+Line 3 no longer has a `!` after `zero`,
+and in line 4 `zeroOne` is inferred to be
+a list of `int`, not `int?`.
+
+<table markdown="1">
+<tr>
+<th>First migration</th>
+<th>Migration with hint</th>
+</tr>
+<tr markdown="1">
+  <td markdown="1">
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0];
+var one = zero! + 1;
+var zeroOne = <int?>[zero, one];
+```
+  </td>
+  <td markdown="1">
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0]/*!*/;
+var one = zero + 1;
+var zeroOne = <int>[zero, one];
+```
+  </td>
+</tr>
+</table>
+
+#### Applying changes
+
+When you like all of the changes
+that the migration tool proposes, click
+**Apply migration** to tell the tool to save the migrated code
+and to update the minimum SDK in your pubspec.
+
+The next step is to [test your code](#step3-test).
```

Comment:
The next step is to ensure that your code has no static errors.  The migration tool may not be able to migrate 100% of your code, particularly if you override it's decisions.  After applying the changes from the tool, you should run the analyzer to ensure that the code now passes static analysis.